python/ray/data/_internal/execution/streaming_executor_state.py

```diff
@@ -0,0 +1,186 @@
+            parent_state = setup_state(parent)
+            op_state.inqueues[i] = parent_state.outqueue
+
+        return op_state
+
+    setup_state(dag)
+
+    # Create the progress bars starting from the first operator to run.
+    i = 0
+    for op_state in list(topology.values())[::-1]:
+        if not isinstance(op_state.op, InputDataBuffer):
+            op_state.initialize_progress_bar(i)
+            i += 1
+
+    # TODO: fill out stats.
+    return topology, DatasetStats(stages={}, parent=None)
+
+
+def process_completed_tasks(topology: Topology) -> None:
+    """Process any newly completed tasks and update operator state."""
+    for op_state in topology.values():
+        op_state.refresh_progress_bar()
+
+    # Update active tasks.
+    active_tasks: Dict[ray.ObjectRef, PhysicalOperator] = {}
+
+    for op in topology:
+        for ref in op.get_work_refs():
+            active_tasks[ref] = op
+
+    # Process completed Ray tasks and notify operators.
+    if active_tasks:
+        completed, _ = ray.wait(
+            list(active_tasks),
+            num_returns=len(active_tasks),
+            fetch_local=False,
+            timeout=0.1,
+        )
+        for ref in completed:
+            op = active_tasks.pop(ref)
+            op.notify_work_completed(ref)
+
+    # Pull any operator outputs into the streaming op state.
+    for op, op_state in topology.items():
+        while op.has_next():
+            op_state.add_output(op.get_next())
+
+    # Call inputs_done() on ops where no more inputs are coming.
+    for op, op_state in topology.items():
+        inputs_done = all(
+            [
+                dep.completed() and not topology[dep].outqueue
+                for dep in op.input_dependencies
+            ]
+        )
+        if inputs_done and not op_state.inputs_done_called:
+            op.inputs_done()
+            op_state.inputs_done_called = True
+
+
+def select_operator_to_run(topology: Topology) -> Optional[PhysicalOperator]:
+    """Select an operator to run, if possible.
+
+    The objective of this function is to maximize the throughput of the overall
+    pipeline, subject to defined memory and parallelism limits.
+
+    This is currently implemented by applying backpressure on operators that are
+    producing outputs faster than they are consuming them `len(outqueue)`, as well as
+    operators with a large number of running tasks `num_active_tasks()`.
+    """
+
+    # TODO: set limits properly based on resources and execution options. This is just
+    # a hard-coded development placeholder.
+    PARALLELISM_LIMIT = 8
+    num_active_tasks = sum(
+        op_state.num_active_tasks() for op_state in topology.values()
+    )
+    if num_active_tasks >= PARALLELISM_LIMIT:
+        return None
+
+    # Equally penalize outqueue length and active tasks for backpressure.
+    pairs = list(topology.items())
+    pairs.sort(key=lambda p: len(p[1].outqueue) + p[1].num_active_tasks())
```

Comment:
Would it make more sense to sort by queue size (desc) and then pick the operator which takes that largest queue as inqueue and so on?

Also do we need to control parallelism per op given we already have a global parallelism control?

Reply:
That doesn't work well since you'd end up running only the first node in the DAG for half the execution. Outqueue is the right one for balancing utilization.

> Also do we need to control parallelism per op given we already have a global parallelism control?

Global control doesn't maximize pipeline throughput. It degrades to bulk execution mode.

Reply:
@ericl Agreed on that formulation for backpressure, but the inqueue size would be a good tie-breaker, right?

Reply:
@ericl Got it, this seems a variant of the max flow problem (https://en.wikipedia.org/wiki/Maximum_flow_problem), the heuristic here makes sense to me!

Reply:
Yup, I think this is probably a very standard online max flow problem. I believe backpressure is a standard heuristic here, but I'm not very familiar with the literature here.

It should be easy to revisit this part of the algorithm in the future.